tools/build.py

```diff
@@ -310,52 +346,102 @@ def Main():
               project_definition.name))
 
       if is_disabled:
-        disabled_packages.append(project_definition.name)
+        disabled_projects.append(project_definition.name)
       else:
         builds.append(project_definition)
 
   if not os.path.exists(options.build_directory):
     os.mkdir(options.build_directory)
 
+  undefined_projects = set(project_names)
+  for disabled_package in disabled_projects:
+    undefined_projects.remove(disabled_package)
+
+  check_configuration = set()
+  failed_builds = set()
+  failed_downloads = set()
+  missing_build_dependencies = set()
+
   current_working_directory = os.getcwd()
   os.chdir(options.build_directory)
 
-  undefined_packages = list(project_names)
-  for disabled_package in disabled_packages:
-    undefined_packages.remove(disabled_package)
+  try:
+    for project_definition in list(builds):
+      if project_names and project_definition.name not in project_names:
+        builds.remove(project_definition)
+        continue
 
-  failed_builds = []
-  for project_definition in builds:
-    if project_names and project_definition.name not in project_names:
-      continue
+      undefined_projects.remove(project_definition.name)
 
-    if undefined_packages:
-      project_index = undefined_packages.index(project_definition.name)
-      del undefined_packages[project_index]
+      if not project_builder.Download(project_definition):
+        builds.remove(project_definition)
 
-    logging.info('Processing: {0:s}'.format(project_definition.name))
+        print('Failed downloading: {0:s}'.format(project_definition.name))
+        failed_downloads.add(project_definition.name)
 
-    # TODO: add support for dokan, bzip2
-    # TODO: setup sqlite in build directory.
-    if not project_builder.Build(project_definition):
-      print('Failed building: {0:s}'.format(project_definition.name))
-      failed_builds.append(project_definition.name)
+    if options.build_target != 'download':
+      for project_definition in list(builds):
```

Comment:
Please log "Conducting prebuild for project: project_name" here, as for building (line 401)

Reply:
Addressed this above, regarding logging the download phase.

Reply:
ack, discussed this.